---
title: "R Notebook"
output: html_notebook
---

# Initialisation

## Packages & functions

```{r, include=FALSE}
source("functions.R")
```

## DATA

```{r, include=FALSE}
RNAseq_diff_gene_expression_analysis <- list()
Gene_set_diff_analysis <- list()
RNAseq <- read.csv("DATA/RNAseq_parsed.csv", row.names = 1, header = T, check.names = F)
Clinical_patient_data <- read.csv("DATA/Clinical_patient_data.csv")
```

## PCA

```{r}
PCA_rnaseq <- PCA(t(RNAseq), ncp = 3)

```


# Differential gene expression

```{r}
Factor_R_OR_NR_B_RUNX1m <- Make_factor(Clinical_patient_data, 
                                                colnames(RNAseq),
                                                c("RUNX1"), 
                                                c("CR", "CRi"), 
                                                "R", 
                                                c("MLFS", "HI", "CRp", "PR"), 
                                                "OR",
                                                c("SD", "PD"), 
                                                "NR")

RNAseq_diff_gene_expression_analysis[["R_OR_NR_B_RUNX1m"]] <- Differential_analysis(Factor_R_OR_NR_B_RUNX1m, RNAseq)

Factor_R_OR_NR_B <- Make_factor(Clinical_patient_data, 
                                                colnames(RNAseq),
                                                0, 
                                                c("CR", "CRi"), 
                                                "R", 
                                                c("MLFS", "HI", "CRp", "PR"), 
                                                "OR",
                                                c("SD", "PD"), 
                                                "NR")

RNAseq_diff_gene_expression_analysis[["R_OR_NR_B"]] <- Differential_analysis(Factor_R_OR_NR_B, RNAseq)

Factor_REL_Baseline <- as.vector(Factor_R_OR_NR_B)
Factor_REL_Baseline[str_detect(Factor_REL_Baseline, "REL")] <- "REL"
Factor_REL_Baseline <- as.factor(Factor_REL_Baseline)

RNAseq_diff_gene_expression_analysis[["REL_Baseline"]] <- Differential_analysis(Factor_REL_Baseline, RNAseq)
```
## Venn diagramm

```{r}
library(VennDiagram)
List_of_DEG <- list()

Vector_DEG_of_interest <- c("NR.B-OR.B", "NR.B-R.B", "OR.B-R.B", "R.B-R.REL")

for (i in Vector_DEG_of_interest){
  List_of_DEG[[paste(i, "UP", sep = "_")]] <- RNAseq_diff_gene_expression_analysis[["R_OR_NR_B"]][[i]] %>% dplyr::filter(logFC > 0.75 & P.Value < 0.1) %>% .$ID
  
  List_of_DEG[[paste(i, "DOWN", sep =  "_")]] <- RNAseq_diff_gene_expression_analysis[["R_OR_NR_B"]][[i]] %>% dplyr::filter(logFC < -0.75 & P.Value < 0.1) %>% .$ID
}

Vector_DEG_of_interest2 <- c("WT.NR.B-WT.OR.B", "WT.NR.B-WT.R.B", "WT.OR.B-WT.R.B", "WT.R.B-WT.R.REL")

for (i in Vector_DEG_of_interest2){
  List_of_DEG[[paste(i, "UP", sep = "_")]] <- RNAseq_diff_gene_expression_analysis[["R_OR_NR_B_RUNX1m"]][[i]] %>% dplyr::filter(logFC > 0.75 & P.Value < 0.1) %>% .$ID
  
  List_of_DEG[[paste(i, "DOWN", sep =  "_")]] <- RNAseq_diff_gene_expression_analysis[["R_OR_NR_B_RUNX1m"]][[i]] %>% dplyr::filter(logFC < -0.75 & P.Value < 0.1) %>% .$ID
}

Vector_DEG_of_interest3 <- c("NR.B-REL", "OR.B-REL", "R.B-REL")

for (i in Vector_DEG_of_interest3){
  List_of_DEG[[paste(i, "UP", sep = "_")]] <- RNAseq_diff_gene_expression_analysis[["REL_Baseline"]][[i]] %>% dplyr::filter(logFC > 0.75 & P.Value < 0.1) %>% .$ID
  
  List_of_DEG[[paste(i, "DOWN", sep =  "_")]] <- RNAseq_diff_gene_expression_analysis[["REL_Baseline"]][[i]] %>% dplyr::filter(logFC < -0.75 & P.Value < 0.1) %>% .$ID
}
```


### No RUNX1m filter

```{r}
Venndiagram_B_df <- list(
  "NR_vs_OR" = c(List_of_DEG$`NR.B-OR.B_UP`, List_of_DEG$`NR.B-OR.B_DOWN`),
  "NR_vs_R" = c(List_of_DEG$`NR.B-R.B_UP`, List_of_DEG$`NR.B-R.B_DOWN`),
  "OR_vs_R" = c(List_of_DEG$`OR.B-R.B_UP`, List_of_DEG$`OR.B-R.B_DOWN`)
)

venn.diagram(Venndiagram_B_df,
             filename = "Results/Figures/Venn_B_DEG.png", 
             imagetype = "png")
```

### RUNX1m filter

```{r}
Venndiagram_B_No_RUNX1_df <- list(
  "NR_vs_OR" = c(List_of_DEG$`WT.NR.B-WT.OR.B_UP`, List_of_DEG$`WT.NR.B-WT.OR.B_DOWN`),
  "NR_vs_R" = c(List_of_DEG$`WT.NR.B-WT.R.B_UP`, List_of_DEG$`WT.NR.B-WT.R.B_DOWN`),
  "OR_vs_R" = c(List_of_DEG$`WT.OR.B-WT.R.B_UP`, List_of_DEG$`WT.OR.B-WT.R.B_DOWN`)
)

venn.diagram(Venndiagram_B_No_RUNX1_df,
             filename = "Results/Figures/Venn_B_DEG_noRUNX1.png", 
             imagetype = "png")
```

## Relapse/Baseline R NR vs R OR vs R

```{r}
Venndiagram_REL_B_df <- list(
  "RBaseline_RRelapse" = c(List_of_DEG$`WT.R.B-WT.R.REL_UP`, List_of_DEG$`WT.R.B-WT.R.REL_DOWN`),
  "NR_vs_R" = c(List_of_DEG$`WT.NR.B-WT.R.B_UP`, List_of_DEG$`WT.NR.B-WT.R.B_DOWN`),
  "OR_vs_R" = c(List_of_DEG$`WT.OR.B-WT.R.B_UP`, List_of_DEG$`WT.OR.B-WT.R.B_DOWN`)
)

venn.diagram(Venndiagram_REL_B_df,
             filename = "Results/Figures/Venndiagram_REL_B_df.png", 
             imagetype = "png")
```


# Gene set enrichment analysis

## Clinical analysis

```{r, include = F}
Gene_set_diff_analysis[["OR_vs_R"]] <- Gene_set_analysis(RNAseq_diff_gene_expression_analysis[["R_OR_NR_B"]][["OR.B-R.B"]])

Gene_set_diff_analysis[["OR_vs_R_without_RUNX1m"]] <- Gene_set_analysis(RNAseq_diff_gene_expression_analysis[["R_OR_NR_B_RUNX1m"]][["WT.OR.B-WT.R.B"]])

Gene_set_diff_analysis[["NR_vs_R"]] <- Gene_set_analysis(RNAseq_diff_gene_expression_analysis[["R_OR_NR_B"]][["NR.B-R.B"]])

Gene_set_diff_analysis[["NR_vs_R_without_RUNX1m"]] <- Gene_set_analysis(RNAseq_diff_gene_expression_analysis[["R_OR_NR_B_RUNX1m"]][["WT.NR.B-WT.R.B"]])
```

## Intersection NR R OR analyses

```{r}
gene_universe <- rownames(RNAseq)

Genes_UP_NR <- intersect(List_of_DEG$`NR.B-OR.B_UP`, List_of_DEG$`NR.B-R.B_UP`)
Genes_DOWN_NR <- intersect(List_of_DEG$`NR.B-OR.B_DOWN`, List_of_DEG$`NR.B-R.B_DOWN`)

Genes_UP_R <- intersect(List_of_DEG$`NR.B-R.B_DOWN`, List_of_DEG$`OR.B-R.B_DOWN`)
Genes_DOWN_R <- intersect(List_of_DEG$`NR.B-R.B_UP`, List_of_DEG$`OR.B-R.B_UP`)

######################## NR ###########################
Gene_set_diff_analysis[["Intersection_analysis"]] <- list()
Gene_set_diff_analysis[["Intersection_analysis"]][["NR_UP"]] <- enrichGO(
    gene = Genes_UP_NR,
    keyType = "SYMBOL",
    OrgDb = "org.Hs.eg.db",
    ont = "ALL",
    pAdjustMethod = "none", 
    universe = gene_universe
  )
Gene_set_diff_analysis[["Intersection_analysis"]][["NR_DOWN"]] <- enrichGO(
    gene = Genes_DOWN_NR,
    keyType = "SYMBOL",
    OrgDb = "org.Hs.eg.db",
    ont = "ALL",
    pAdjustMethod = "none", 
    universe = gene_universe
  )

#################### R ########################
Gene_set_diff_analysis[["Intersection_analysis"]][["R_UP"]] <- enrichGO(
    gene = Genes_UP_R,
    keyType = "SYMBOL",
    OrgDb = "org.Hs.eg.db",
    ont = "ALL",
    pAdjustMethod = "none", 
    universe = gene_universe
  )
Gene_set_diff_analysis[["Intersection_analysis"]][["R_DOWN"]] <- enrichGO(
    gene = Genes_DOWN_R,
    keyType = "SYMBOL",
    OrgDb = "org.Hs.eg.db",
    ont = "ALL",
    pAdjustMethod = "none", 
    universe = gene_universe
  )

rm(Genes_UP_NR, Genes_DOWN_NR, Genes_UP_R, Genes_DOWN_R)
```

## Intersection Relapse R analyses

```{r}
Genes_UP_R_B_UP <- intersect(List_of_DEG$`NR.B-R.B_DOWN`, List_of_DEG$`R.B-R.REL_UP`)
Genes_UP_R_B_DOWN <- intersect(List_of_DEG$`NR.B-R.B_UP`, List_of_DEG$`R.B-R.REL_DOWN`)

Gene_set_diff_analysis[["Intersection_analysis"]][["REL_R_B_UP"]] <- enrichGO(
    gene = Genes_UP_R_B_UP,
    keyType = "SYMBOL",
    OrgDb = "org.Hs.eg.db",
    ont = "ALL",
    pAdjustMethod = "none", 
    universe = gene_universe
  )
Gene_set_diff_analysis[["Intersection_analysis"]][["REL_R_B_DOWN"]] <- enrichGO(
    gene = Genes_UP_R_B_DOWN,
    keyType = "SYMBOL",
    OrgDb = "org.Hs.eg.db",
    ont = "ALL",
    pAdjustMethod = "none", 
    universe = gene_universe
  )

rm(Genes_UP_R_B_UP, Genes_UP_R_B_DOWN)
```

## Intersection Relapse vs OR analyses

```{r}
Genes_UP_OR_B_UP <- intersect(List_of_DEG$`NR.B-OR.B_DOWN`, List_of_DEG$`OR.B-REL_UP`)
Genes_DOWN_OR_B_DOWN <- intersect(List_of_DEG$`NR.B-OR.B_UP`, List_of_DEG$`OR.B-REL_DOWN`)

Gene_set_diff_analysis[["Intersection_analysis"]][["REL_OR_B_UP"]] <- enrichGO(
    gene = Genes_UP_OR_B_UP,
    keyType = "SYMBOL",
    OrgDb = "org.Hs.eg.db",
    ont = "ALL",
    pAdjustMethod = "none", 
    universe = gene_universe
  )
Gene_set_diff_analysis[["Intersection_analysis"]][["REL_OR_B_DOWN"]] <- enrichGO(
    gene = Genes_DOWN_OR_B_DOWN,
    keyType = "SYMBOL",
    OrgDb = "org.Hs.eg.db",
    ont = "ALL",
    pAdjustMethod = "none", 
    universe = gene_universe
  )

rm(Genes_UP_OR_B_UP, Genes_DOWN_OR_B_DOWN)
```



## Relapse versus Baseline R

```{r}
Gene_set_diff_analysis[["REL_B.R"]] <- Gene_set_analysis(RNAseq_diff_gene_expression_analysis[["REL_Baseline"]][["R.B-REL"]])
```

# MS Viper analyses

```{r}
data(dorothea_hs, package = "dorothea") 
regulons = dorothea_hs %>%
  filter(confidence %in% c("A", "B"))

ref_R_B <- Factor_R_OR_NR_B == "R.B"
ref_OR_B <- Factor_R_OR_NR_B == "OR.B"
ref_NR_B <- Factor_R_OR_NR_B == "NR.B"
ref_R_REL <- Factor_R_OR_NR_B == "R.REL" | Factor_R_OR_NR_B == "OR.REL" | Factor_R_OR_NR_B == "NR.REL" 

R_OR_msviper <- Do_MS_viper_analysis(RNAseq, regulons, ref_R_B, "R", ref_OR_B, "OR")

R_NR_msviper <- Do_MS_viper_analysis(RNAseq, regulons, ref_R_B, "R", ref_NR_B, "NR")

OR_NR_msviper <- Do_MS_viper_analysis(RNAseq, regulons, ref_OR_B, "OR", ref_NR_B, "NR")

R_B_REL_msviper <- Do_MS_viper_analysis(RNAseq, regulons, ref_R_B, "R", ref_R_REL, "R_REL")

OR_B_REL_msviper <- Do_MS_viper_analysis(RNAseq, regulons, ref_OR_B, "R", ref_R_REL, "OR_REL")
```

## Intersection 

```{r}
TF_UP_NR_R <- R_NR_msviper[["MSVIPER"]]$mrs_table %>% dplyr::filter(pval < 0.1 & nes > 0) %>% .$TF
TF_UP_REL_R <- R_B_REL_msviper[["MSVIPER"]]$mrs_table %>% dplyr::filter(pval < 0.1 & nes > 0) %>% .$TF

TF_DOWN_NR_R <- R_NR_msviper[["MSVIPER"]]$mrs_table %>% dplyr::filter(pval < 0.1 & nes < 0) %>% .$TF
TF_DOWN_REL_R <- R_B_REL_msviper[["MSVIPER"]]$mrs_table %>% dplyr::filter(pval < 0.1 & nes < 0) %>% .$TF

TF_UP_NR_and_REL_vs_R <- c(intersect(TF_UP_NR_R, TF_UP_REL_R), intersect(TF_DOWN_NR_R,TF_DOWN_REL_R))



TF_UP_OR_R <- R_OR_msviper[["MSVIPER"]]$mrs_table %>% dplyr::filter(pval < 0.1 & nes > 0) %>% .$TF
TF_UP_REL_OR <- OR_B_REL_msviper[["MSVIPER"]]$mrs_table %>% dplyr::filter(pval < 0.1 & nes > 0) %>% .$TF

TF_DOWN_OR_R <- R_OR_msviper[["MSVIPER"]]$mrs_table %>% dplyr::filter(pval < 0.1 & nes < 0) %>% .$TF
TF_DOWN_REL_OR <- OR_B_REL_msviper[["MSVIPER"]]$mrs_table %>% dplyr::filter(pval < 0.1 & nes < 0) %>% .$TF

TF_UP_NR_and_REL_vs_OR <- c(intersect(TF_UP_OR_R, TF_UP_REL_R), intersect(TF_DOWN_OR_R, TF_DOWN_REL_R))

rm(TF_UP_NR_R, TF_UP_REL_R, TF_DOWN_NR_R, TF_DOWN_REL_R, TF_UP_OR_R, TF_UP_REL_OR, TF_DOWN_OR_R, TF_DOWN_REL_OR)
```

# Preparing Combining Networks

```{r}
PPI_Network <- read.csv("DATA/FIsInGene_020720_with_annotations.tsv", sep = "\t")

# attribute value	explanation
# ECrel	enzyme-enzyme relation, indicating two enzymes catalyzing successive reaction steps
# PPrel	protein-protein interaction, such as binding and modification
# GErel	gene expression interaction, indicating relation of transcription factor and target gene product
# PCrel	protein-compound interaction
# maplink	link to another map

PPI_Network$Edge_ID <- paste(PPI_Network$Gene1, PPI_Network$Gene2, sep = ".")

Combined_Networks <- regulons
Combined_Networks$Edge_ID <- paste(Combined_Networks$tf, Combined_Networks$target, sep = ".")

Combined_Networks <- merge(Combined_Networks, PPI_Network, by = "Edge_ID", all = T)
Combined_Networks$Gene1 <- ifelse(is.na(Combined_Networks$Gene1), Combined_Networks$tf, Combined_Networks$Gene1)
Combined_Networks$Gene2 <- ifelse(is.na(Combined_Networks$Gene2), Combined_Networks$target, Combined_Networks$Gene2)
Combined_Networks <- Combined_Networks[,c(6, 7, 3, 5, 8:10)]
colnames(Combined_Networks) <- c("Gene_IN", "Gene_Out", "confidence", "Tf_Regulation", "PPI_Annotation", "Direction_PPI", "Score_PPI")

Combined_Networks$Tf_Regulation <- ifelse(is.na(Combined_Networks$Tf_Regulation), 0, Combined_Networks$Tf_Regulation)
Combined_Networks$confidence <- ifelse(is.na(Combined_Networks$confidence), "", Combined_Networks$confidence)
Combined_Networks$PPI_Annotation <- ifelse(is.na(Combined_Networks$PPI_Annotation), "", Combined_Networks$PPI_Annotation)
Combined_Networks$Direction_PPI <- ifelse(is.na(Combined_Networks$Direction_PPI), "", Combined_Networks$Direction_PPI)
Combined_Networks$Score_PPI <- ifelse(is.na(Combined_Networks$Score_PPI), 0.0, Combined_Networks$Score_PPI)

GErel_annotations <- c("GErel", "regulat", "repress", "expression")
PPrel_annotations <-c("activat", "binding", "catalyze", "complex", "compound", "phosphyr", "dissocia", "effect", "glycol", "interact", "methyla", "PPrel", "state", "ubiqui")
ECrel_annotation <- c("ECrel", "reaction")

tmp <- sapply(Combined_Networks$PPI_Annotation, function(x){
  GE_test <- sapply(GErel_annotations, function(GE){
    (str_detect(x, GE))*1
  }) %>% sum() > 0
  EC_test <- sapply(ECrel_annotation, function(EC){
    (str_detect(x, EC))*1
  }) %>% sum() > 0
  PP_test <- sapply(PPrel_annotations, function(PP){
    (str_detect(x, PP))*1
  }) %>% sum() > 0
  predicted_test <- str_detect(x, "predicted")
  ifelse(GE_test, "GErel", ifelse(EC_test, "ECrel", ifelse(PP_test, "PPrel", ifelse(predicted_test, "predicted", "other"))))
})

Combined_Networks$Cleaned_PPI_annotation <- tmp
```




# Saving

## Tables

```{r}
table_dir <- "Results/Tables/"
```


### DEG

```{r}
write.csv(RNAseq_diff_gene_expression_analysis[["R_OR_NR_B"]][["NR.B-R.B"]], paste0(table_dir, "DEG_NR_vs_R.csv"), row.names = T)
write.csv(RNAseq_diff_gene_expression_analysis[["REL_Baseline"]][["R.B-REL"]], paste0(table_dir, "DEG_R_vs_REL.csv"), row.names = T)

Double_DEG_NR_R_REL <- merge(RNAseq_diff_gene_expression_analysis[["R_OR_NR_B"]][["NR.B-R.B"]], RNAseq_diff_gene_expression_analysis[["REL_Baseline"]][["R.B-REL"]], by="ID") %>%
  merge(RNAseq_diff_gene_expression_analysis[["R_OR_NR_B"]][["OR.B-R.B"]], by = "ID") %>%
  merge(RNAseq_diff_gene_expression_analysis[["R_OR_NR_B"]][["NR.B-OR.B"]], by = "ID")
Double_DEG_NR_R_REL <- Double_DEG_NR_R_REL[, c(1, 2, 5, 8, 11, 14, 17, 20, 23)]
colnames(Double_DEG_NR_R_REL) <- c("ID", "NR_R_logFC", "NR_R_P.Value", "R_REL_logFC", "R_REL_P.Value", "OR_R_logFC", "OR_R_P.Value", "NR_OR_logFC", "NR_OR_P.Value")

write.csv(Double_DEG_NR_R_REL, paste0(table_dir, "Double_DEG_NR_R_REL.csv"), row.names = F)
```

### MS_viper

```{r}
write.csv(R_NR_msviper[["MSVIPER"]]$mrs_table, "Results/Tables/NR_vs_R_msviper.csv", row.names = F)
write.csv(R_OR_msviper[["MSVIPER"]]$mrs_table, "Results/Tables/OR_vs_R_msviper.csv", row.names = F)
write.csv(R_B_REL_msviper[["MSVIPER"]]$mrs_table, "Results/Tables/REL_vs_R_msviper.csv", row.names = F)

Double_MS_Viper_NR_R_REL <- merge(R_NR_msviper[["MSVIPER"]]$mrs_table, R_B_REL_msviper[["MSVIPER"]]$mrs_table, by = "TF") %>%
  merge(R_OR_msviper[["MSVIPER"]]$mrs_table, by = "TF") %>%
  merge(OR_NR_msviper[["MSVIPER"]]$mrs_table, by = "TF")

Double_MS_Viper_NR_R_REL <- Double_MS_Viper_NR_R_REL[, c(1, 3, 4, 7, 8, 11, 12, 15, 16)]
colnames(Double_MS_Viper_NR_R_REL) <- c("ID", "NR_R_nes", "NR_R_pval", "REL_R_nes", "REL_R_pval", "OR_R_nes", "OR_R_pval", "NR_OR_nes", "NR_OR_pval")

write.csv(Double_MS_Viper_NR_R_REL, paste0(table_dir, "Double_MS_Viper_NR_R_REL.csv"), row.names = F)
```

### Networks

```{r}
write.table(Combined_Networks, "Results/Tables/Combined_Networks.tsv", row.names = F, sep = "\t", quote = FALSE)
```



## Figures

```{r}
fig_dir <- "Results/Figures/"
```

### GO

```{r}
png("Results/Figures/GO_NR_R.png", width = 720, height = 720*9/16)
dotplot(Gene_set_diff_analysis[["NR_vs_R"]], showCategory=10, split=".sign") + facet_grid(.~.sign)
dev.off()

png("Results/Figures/GO_Rel_R.png", width = 720, height = 720*9/16)
dotplot(Gene_set_diff_analysis[["REL_B.R"]], showCategory=10, split=".sign") + facet_grid(.~.sign)
dev.off()



png("Results/Figures/GO_NR_UP.png")
emapplot(Gene_set_diff_analysis[["Intersection_analysis"]][["NR_UP"]], color = "qvalue", size = "Count")
dev.off()
png("Results/Figures/GO_NR_DOWN.png")
emapplot(Gene_set_diff_analysis[["Intersection_analysis"]][["NR_DOWN"]], color = "qvalue", size = "Count")
dev.off()

png("Results/Figures/GO_R_UP.png")
emapplot(Gene_set_diff_analysis[["Intersection_analysis"]][["R_UP"]], color = "qvalue", size = "Count")
dev.off()

png("Results/Figures/GO_REL_R_B_DOWN.png")
emapplot(Gene_set_diff_analysis[["Intersection_analysis"]][["REL_R_B_DOWN"]], color = "qvalue", size = "Count")
dev.off()
```

### MS_VIPER

```{r}
png(paste0(fig_dir, "MS_Viper_NR_R.png"))
plot(R_NR_msviper[["MSVIPER"]]$mrs)
dev.off()
png(paste0(fig_dir, "MS_Viper_OR_R.png"))
plot(R_OR_msviper[["MSVIPER"]]$mrs)
dev.off()
png(paste0(fig_dir, "MS_Viper_REL_R.png"))
plot(R_B_REL_msviper[["MSVIPER"]]$mrs)
dev.off()
```

# Signatures

```{r}
message("PID signature")
PID_VDR_RXRA_PATHWAy <- read.csv("../../REVISION_VD_IDH_Alexis/Autres Signatures géniques/PID_RXR_VDR_PATHWAY - Feuille.csv", header = F)$V1
head(PID_VDR_RXRA_PATHWAy, 30)

message("Early signature")
Early_signature <- read.csv("../../REVISION_VD_IDH_Alexis/Results/Intersect_0.75_0.05_4h.csv", row.names = 1)
head(Early_signature, 30)

message("Late signature")
Late_signature <- read.csv("../../REVISION_VD_IDH_Alexis/Results/Intersect_0.75_0.05_24h.csv", row.names = 1)
head(Late_signature, 30)

message("WP_VDR_PATHWAY")
WP_VITAMIN_D_RECEPTOR_PATHWAY <- read.csv("../../REVISION_VD_IDH_Alexis/Autres Signatures géniques/VITAMIN_D_RECEPTOR_PATHWAY.txt", skip = 2, header = F)$V1
head(WP_VITAMIN_D_RECEPTOR_PATHWAY, 30)
```

```{r}
volcanoplot_gene_expression <- function(DATA, title, xlim = 10, ylim = 5, pvalue_cutoff = 0.1, logFCcutoff = 0.75){
  enhancedvolcano_data1 <- data.frame(logFC = DATA[, "logFC"],
                                      pvalue = DATA[, "P.Value"],
                                      gene = DATA[, "ID"])
  
  EnhancedVolcano(toptable = enhancedvolcano_data1, 
                  lab = enhancedvolcano_data1$gene, 
                  x = "logFC", 
                  y = "pvalue",
                  FCcutoff = logFCcutoff,
                  pCutoff = pvalue_cutoff,
                  title = title,
                  subtitle = NA,
                  legendPosition = "right",
                  subtitleLabSize = 0,
                  legendLabSize = 10,
                  ylim = c(0, ylim),
                  xlim = c(-xlim, xlim)
  )
}

```

## EARLY

### NR vs R

```{r}
volcanoplot_gene_expression(RNAseq_diff_gene_expression_analysis[["R_OR_NR_B"]][["NR.B-R.B"]][Early_signature,], "Early VDR signature", ylim = 10, xlim = 5)
```

## LATE

### NR vs R

```{r}
volcanoplot_gene_expression(RNAseq_diff_gene_expression_analysis[["R_OR_NR_B"]][["NR.B-R.B"]][Late_signature,], "Late VDR Signature", ylim = 10, xlim = 5)
```

## WP_VITAMIN_D_RECEPTOR_PATHWAY

### NR vs R

```{r}
volcanoplot_gene_expression(RNAseq_diff_gene_expression_analysis[["R_OR_NR_B"]][["NR.B-R.B"]][WP_VITAMIN_D_RECEPTOR_PATHWAY,], "WP VDR Signature", ylim = 10, xlim = 5)
```

# Clustering based on pathways

```{r}
make_heatmap_pathway <- function(data, signature, title, anno){
  tmp <- data[signature,] %>% as.matrix()
  corr <- rcorr(tmp, type = "spearman")$r
  pheatmap(corr, main = title, annotation_col = anno)
}

RNAseq_Baseline <- RNAseq[,which(colnames(RNAseq) %in% Clinical_patient_data$Baseline_RNAseq_data)]
Response <- Clinical_patient_data[which(Clinical_patient_data$Baseline_RNAseq_data %in% colnames(RNAseq_Baseline)),c(5,9)] %>% unique()
rownames(Response) <- Response$Baseline_RNAseq_data
Response$Best_response <- ifelse(Response$Best_response %in% c("CR", "CRi"), "R", ifelse(Response$Best_response %in% c("SD", "PD"), "NR", "OR"))

annotation_heatmap <- data.frame(Responses = Response[colnames(RNAseq_Baseline),"Best_response"])
rownames(annotation_heatmap) <- colnames(RNAseq_Baseline)


make_heatmap_pathway(RNAseq_Baseline, PID_VDR_RXRA_PATHWAy, "PID_pathway", annotation_heatmap)
make_heatmap_pathway(RNAseq_Baseline, Early_signature, "Early_signature", annotation_heatmap)
make_heatmap_pathway(RNAseq_Baseline, Late_signature, "Late_signature", annotation_heatmap)
make_heatmap_pathway(RNAseq_Baseline, WP_VITAMIN_D_RECEPTOR_PATHWAY, "WP_VITAMIN_D_RECEPTOR_PATHWAY", annotation_heatmap)

```

## EARLY

### REL vs R

```{r}
volcanoplot_gene_expression(RNAseq_diff_gene_expression_analysis[["REL_Baseline"]][["R.B-REL"]][Early_signature,], "Early VDR signature", ylim = 10, xlim = 5)
```

## LATE

### NR vs R

```{r}
volcanoplot_gene_expression(RNAseq_diff_gene_expression_analysis[["REL_Baseline"]][["R.B-REL"]][Late_signature,], "Late VDR Signature", ylim = 10, xlim = 5)
```

## WP_VITAMIN_D_RECEPTOR_PATHWAY

### NR vs R

```{r}
volcanoplot_gene_expression(RNAseq_diff_gene_expression_analysis[["REL_Baseline"]][["R.B-REL"]][WP_VITAMIN_D_RECEPTOR_PATHWAY,], "WP VDR Signature", ylim = 10, xlim = 5)
```

# Clustering based on pathways

```{r}
make_heatmap_pathway <- function(data, signature, title, anno){
  tmp <- data[signature,] %>% as.matrix()
  corr <- rcorr(tmp, type = "pearson")$r
  pheatmap(corr, main = title, annotation_col = anno)
}

RNAseq_R_REL <- RNAseq[,Factor_REL_Baseline == "REL" | Factor_REL_Baseline == "R.B"]

annotation_heatmap <- data.frame(Responses = Factor_REL_Baseline[Factor_REL_Baseline == "REL" | Factor_REL_Baseline == "R.B"])
rownames(annotation_heatmap) <- colnames(RNAseq_R_REL)


make_heatmap_pathway(RNAseq_R_REL, PID_VDR_RXRA_PATHWAy, "PID_pathway", annotation_heatmap)
make_heatmap_pathway(RNAseq_R_REL, Early_signature, "Early_signature", annotation_heatmap)
make_heatmap_pathway(RNAseq_R_REL, Late_signature, "Late_signature", annotation_heatmap)
make_heatmap_pathway(RNAseq_R_REL, WP_VITAMIN_D_RECEPTOR_PATHWAY, "WP_VITAMIN_D_RECEPTOR_PATHWAY", annotation_heatmap)

```

# GSEA analyses

```{r}
DEG_tmp <- dplyr::filter(RNAseq_diff_gene_expression_analysis$R_OR_NR_B$`NR.B-R.B`, P.Value < 0.1)

gene_list_1 <- DEG_tmp$logFC
names(gene_list_1) <- DEG_tmp$ID
gene_list_1 = sort(gene_list_1, decreasing = TRUE)

gene_list_1 = gene_list_1[!duplicated(names(gene_list_1))]
head(gene_list_1)
```

```{r}
myGO = fgsea::gmtPathways("~/R/x86_64-pc-linux-gnu-library/4.0/fgsea/extdata/msigdb.v7.4.symbols.gmt")
```

```{r}
fgsea::plotEnrichment(myGO$WP_VITAMIN_D_RECEPTOR_PATHWAY, gene_list_1) + labs(title = "WP signature VITAMIN D RECEPTOR PATHWAY TCGA")
fgsea::plotEnrichment(myGO$PID_RXR_VDR_PATHWAY, gene_list_1) + labs(title = "PID RXRA VITAMIN D RECEPTOR PATHWAY TCGA")

Early_UP <- Early_signature[which(Early_signature$log2FoldChange > 0),] %>% rownames()
Early_Down <- Early_signature[which(Early_signature$log2FoldChange < 0),] %>% rownames()
Late_UP <- Late_signature[which(Late_signature$log2FoldChange > 0),] %>% rownames()
Late_Down <- Late_signature[which(Late_signature$log2FoldChange < 0),] %>% rownames()

fgsea::plotEnrichment(rownames(Early_signature), gene_list_1) + labs(title = "Early VDR signature")
fgsea::plotEnrichment(rownames(Late_signature), gene_list_1) + labs(title = "Late VDR signature")

# fgsea::plotEnrichment(Early_UP, gene_list_1) + labs(title = "Early_up VDR signature")
# fgsea::plotEnrichment(Late_UP, gene_list_1) + labs(title = "Late_up VDR signature")
# 
# fgsea::plotEnrichment(Early_Down, gene_list_1) + labs(title = "Early_down VDR signature")
# fgsea::plotEnrichment(Late_Down, gene_list_1) + labs(title = "Late VDR_down signature")


```


```{r}
DEG_tmp2 <- dplyr::filter(RNAseq_diff_gene_expression_analysis$REL_Baseline$`R.B-REL`, P.Value < 0.1)

gene_list_2 <- DEG_tmp2$logFC
names(gene_list_2) <- DEG_tmp2$ID
gene_list_2 = sort(gene_list_2, decreasing = TRUE)

gene_list_2 = gene_list_2[!duplicated(names(gene_list_2))]
head(gene_list_2)
```

```{r}
fgsea::plotEnrichment(myGO$WP_VITAMIN_D_RECEPTOR_PATHWAY, gene_list_2) + labs(title = "WP signature VITAMIN D RECEPTOR PATHWAY TCGA")
fgsea::plotEnrichment(myGO$PID_RXR_VDR_PATHWAY, gene_list_2) + labs(title = "PID RXRA VITAMIN D RECEPTOR PATHWAY TCGA")

fgsea::plotEnrichment(Early_signature, gene_list_2) + labs(title = "Early VDR signature")
fgsea::plotEnrichment(Late_signature, gene_list_2) + labs(title = "Late VDR signature")
```

# TF activities

```{r}
tf_activities <- run_viper(RNAseq, regulons, 
                           options =  list(method = "scale", minsize = 4, eset.filter = FALSE, cores = 1, verbose = FALSE))
```

# Focus on one gene

```{r}
Focus_on_one_gene <- function(RNAseq, Gene, Comparison_A, Comparison_A_name, Comparison_B, Comparison_B_name, phenotype, tfs){
  df_tf <- tfs[rownames(tfs) == Gene, phenotype %in% c(Comparison_A, Comparison_B)]
  df <- RNAseq[str_detect(rownames(RNAseq), Gene), phenotype %in% c(Comparison_A, Comparison_B)]
  phenotype <- phenotype[phenotype %in% c(Comparison_A, Comparison_B)]
  pheno <- ifelse(phenotype == Comparison_A, Comparison_A_name, Comparison_B_name)
  df <- t(df) %>% as.data.frame()
  df_tf <- df_tf %>% as.data.frame()
  df$Phenotype <- pheno
  df_tf$Phenotype <- pheno
  
  df[,1] <- as.numeric(df[,1])
  df_tf[,1] <- as.numeric(df_tf[,1])
  colnames(df_tf)[1] <- "Gene"
  colnames(df)[1] <- "Gene"
  res <- list("Expr" = df,
              "TF" = df_tf)
  res
}
```


```{r}
Make_gene_expr_tf_activity_boxplots <- function(RNAseq, TF_to_focus, Comparison_A, Comparison_A_name, Comparison_B, Comparison_B_name, Phenotype, tf_activities){
  Phenotype_of_interest <- Phenotype

  Data_on_the_gene <- Focus_on_one_gene(RNAseq, TF_to_focus, Comparison_A, Comparison_A_name, Comparison_B, Comparison_B_name, Phenotype, tf_activities)
  par(mfrow=c(1,2))
  expr_plot <- ggplot(Data_on_the_gene$Expr, aes(x=Phenotype, y = Gene, fill=Phenotype))+
    geom_boxplot() +
    geom_jitter(Data_on_the_gene$Expr, inherit.aes = FALSE, mapping = aes(y = Gene, x = Phenotype), width = 0.25, alpha = 0.5, colour = "darkred")+

    geom_signif(
      comparisons = list(c(Comparison_A_name, Comparison_B_name))
    ) +
    ggtitle(paste0(TF_to_focus, " expression voom normalized"))+
    ylab("Voom expression")
  # dev.off()
  
  # png(paste0(TF_to_focus, "_NR_R_TF_boxplot.png"))
  tf_plot <- ggplot(Data_on_the_gene$TF, aes(x=Phenotype, y = Gene, fill=Phenotype))+
    geom_boxplot() +
    geom_jitter(Data_on_the_gene$TF, inherit.aes = FALSE, mapping = aes(y = Gene, x = Phenotype), width = 0.25, alpha = 0.5, colour = "darkred")+
    geom_signif(
      comparisons = list(c(Comparison_A_name, Comparison_B_name))
    ) +
    ggtitle(paste0(TF_to_focus, " TF activity"))+
    ylab("Tf activity")
  plot_grid(expr_plot, tf_plot, labels=c("A", "B"), ncol = 2, nrow = 1)
  ggsave(paste0(TF_to_focus, "_", Comparison_A_name, "_", Comparison_B_name, "_boxplots.png"))

}
```

```{r}
Make_gene_expr_tf_activity_boxplots(RNAseq, "RUNX1", "R.B", "Responder", "NR.B", "Bad_responder", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "RELA",  "R.B", "Responder", "NR.B", "Bad_responder", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "CEBPA",  "R.B", "Responder", "NR.B", "Bad_responder", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "HIF1A",  "R.B", "Responder", "NR.B", "Bad_responder", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "MYC",  "R.B", "Responder", "NR.B", "Bad_responder", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "SMAD3",  "R.B", "Responder", "NR.B", "Bad_responder", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "REL",  "R.B", "Responder", "NR.B", "Bad_responder", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "JUN",  "R.B", "Responder", "NR.B", "Bad_responder", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "CREB1",  "R.B", "Responder", "NR.B", "Bad_responder", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "FOXO3",  "R.B", "Responder", "NR.B", "Bad_responder", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "RXRA",  "R.B", "Responder", "NR.B", "Bad_responder", Factor_R_OR_NR_B, tf_activities)

```

```{r}
Make_gene_expr_tf_activity_boxplots(RNAseq, "RUNX1", "R.B", "Baseline", c("OR.REL", "R.REL"), "Relapse", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "RELA", "R.B", "Baseline", c("OR.REL", "R.REL"), "Relapse", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "CEBPA", "R.B", "Baseline", c("OR.REL", "R.REL"), "Relapse", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "HIF1A", "R.B", "Baseline", c("OR.REL", "R.REL"), "Relapse", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "MYC", "R.B", "Baseline", c("OR.REL", "R.REL"), "Relapse", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "SMAD3", "R.B", "Baseline", c("OR.REL", "R.REL"), "Relapse",Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "REL", "R.B", "Baseline", c("OR.REL", "R.REL"), "Relapse", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "JUN", "R.B", "Baseline", c("OR.REL", "R.REL"), "Relapse", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "CREB1", "R.B", "Baseline", c("OR.REL", "R.REL"), "Relapse", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "FOXO3", "R.B", "Baseline", c("OR.REL", "R.REL"), "Relapse", Factor_R_OR_NR_B, tf_activities)
Make_gene_expr_tf_activity_boxplots(RNAseq, "RXRA", "R.B", "Baseline", c("OR.REL", "R.REL"), "Relapse", Factor_R_OR_NR_B, tf_activities)

```

```{r}
library(viper)
data(dorothea_hs, package = "dorothea") 
regulons = dorothea_hs %>%
  filter(confidence %in% c("A", "B"))
```


```{r}
tf_activities_TCGA <- run_viper(TCGA_norm_data, regulons, 
                           options =  list(method = "scale", minsize = 4, eset.filter = FALSE, cores = 1, verbose = FALSE))

Make_factor_TCGA <- function(Samplesheet){
  IDH1_typeof_mutations <- c("p.R132C", "p.R132G", "p.R132H", "p.R132S")
  IDH2_typeof_mutations <- c("p.R140L", "p.R140Q", "p.R172K")
  IDH1_factor <- ifelse(Samplesheet["IDH1",] %in% IDH1_typeof_mutations, "IDH1m", "")
  IDH2_factor <- ifelse(Samplesheet["IDH2",] %in% IDH2_typeof_mutations, "IDH2m", "")
  IDH_factor <- paste(IDH1_factor, IDH2_factor, sep = ".")
  final_factor <- ifelse(IDH1_factor == "IDH1m" | IDH2_factor == "IDH2m", "IDHm", "IDHwt")
  res <- list("IDHm" = final_factor, "IDH12m" = IDH_factor)
  return(res)
}

TCGA_IDH_factor <- Make_factor_TCGA(TCGA_clinicals)

Make_gene_expr_tf_activity_boxplots(TCGA_norm_data, "RUNX1", "IDHm", "IDHm", "IDHwt", "IDHwt", TCGA_IDH_factor$IDHm, tf_activities_TCGA)
Make_gene_expr_tf_activity_boxplots(TCGA_norm_data, "TFAP2A", "IDHm", "IDHm", "IDHwt", "IDHwt", TCGA_IDH_factor$IDHm, tf_activities_TCGA)
Make_gene_expr_tf_activity_boxplots(TCGA_norm_data, "TFAP2C", "IDHm", "IDHm", "IDHwt", "IDHwt", TCGA_IDH_factor$IDHm, tf_activities_TCGA)
Make_gene_expr_tf_activity_boxplots(TCGA_norm_data, "REST", "IDHm", "IDHm", "IDHwt", "IDHwt", TCGA_IDH_factor$IDHm, tf_activities_TCGA)

```

