---
title: "R Notebook"
output: html_notebook
---

```{r, include=FALSE}
source("functions.R")

RNAseq_diff_gene_expression_analysis <- list()
Gene_set_diff_analysis <- list()
RNAseq <- read.csv("DATA/RNAseq_parsed.csv", row.names = 1, header = T, check.names = F)
Clinical_patient_data <- read.csv("DATA/Clinical_patient_data.csv")
```

# Differential gene expression

```{r}
Factor_R_OR_NR_Baseline_RUNX1m <- Make_factor(Clinical_patient_data, 
                                                colnames(RNAseq),
                                                c("RUNX1"), 
                                                c("CR", "CRi"), 
                                                "Responder", 
                                                c("MLFS", "HI", "CRp", "PR"), 
                                                "Overall_responder",
                                                c("SD", "PD"), 
                                                "Non_responder")

RNAseq_diff_gene_expression_analysis[["R_OR_NR_Baseline_RUNX1m"]] <- Differential_analysis(Factor_R_OR_NR_Baseline_RUNX1m, RNAseq)

Factor_R_OR_NR_Baseline <- Make_factor(Clinical_patient_data, 
                                                colnames(RNAseq),
                                                0, 
                                                c("CR", "CRi"), 
                                                "Responder", 
                                                c("MLFS", "HI", "CRp", "PR"), 
                                                "Overall_responder",
                                                c("SD", "PD"), 
                                                "Non_responder")

RNAseq_diff_gene_expression_analysis[["R_OR_NR_Baseline"]] <- Differential_analysis(Factor_R_OR_NR_Baseline, RNAseq)


```

# Gene set enrichment analysis

```{r, include = F}
Gene_set_diff_analysis[["OR_vs_R"]] <- Gene_set_analysis(RNAseq_diff_gene_expression_analysis[["R_OR_NR_Baseline"]][["Overall_responder.Baseline-Responder.Baseline"]])

Gene_set_diff_analysis[["OR_vs_R_without_RUNX1m"]] <- Gene_set_analysis(RNAseq_diff_gene_expression_analysis[["R_OR_NR_Baseline_RUNX1m"]][["WT.Overall_responder.Baseline-WT.Responder.Baseline"]])

Gene_set_diff_analysis[["NR_vs_R"]] <- Gene_set_analysis(RNAseq_diff_gene_expression_analysis[["R_OR_NR_Baseline"]][["Non_responder.Baseline-Responder.Baseline"]])

Gene_set_diff_analysis[["NR_vs_R_without_RUNX1m"]] <- Gene_set_analysis(RNAseq_diff_gene_expression_analysis[["R_OR_NR_Baseline_RUNX1m"]][["WT.Non_responder.Baseline-WT.Responder.Baseline"]])
```

# MS Viper analyses

```{r}
data(dorothea_hs, package = "dorothea") 
regulons = dorothea_hs %>%
  filter(confidence %in% c("A", "B"))

ref_R_Baseline <- Factor_R_OR_NR_Baseline == "Responder.Baseline"
ref_OR_Baseline <- Factor_R_OR_NR_Baseline == "Overall_responder.Baseline"
ref_NR_Baseline <- Factor_R_OR_NR_Baseline == "Non_responder.Baseline"

R_OR_msViper <- run_msviper(RNAseq, regulons, use_aracne = T, ref = ref_R_Baseline, ref_name = "R", treat = ref_OR_Baseline, treat_name = "OR", minsize = 4, ges.filter = T)

R_OR_msViper_combinatory <- msviperCombinatorial(R_OR_msViper$mrs, regulators = 25, verbose = FALSE)
R_OR_msViper_synergy <- msviperSynergy(R_OR_msViper_combinatory, verbose = FALSE)

R_NR_msViper <- run_msviper(RNAseq, regulons, use_aracne = T, ref = ref_R_Baseline, ref_name = "R", treat = ref_NR_Baseline, treat_name = "NR", minsize = 4, ges.filter = T)

R_NR_msViper_combinatory <- msviperCombinatorial(R_NR_msViper$mrs, regulators = 25, verbose = FALSE)
R_NR_msViper_synergy <- msviperSynergy(R_NR_msViper_combinatory, verbose = FALSE)

ref_R_Baseline_no_RUNX1m <- Factor_R_OR_NR_Baseline_RUNX1m == "WT.Responder.Baseline"
ref_OR_Baseline_no_RUNX1m <- Factor_R_OR_NR_Baseline_RUNX1m == "WT.Overall_responder.Baseline"
ref_NR_Baseline_no_RUNX1m <- Factor_R_OR_NR_Baseline_RUNX1m == "WT.Non_responder.Baseline"

R_OR_no_RUNX1m_msViper <- run_msviper(RNAseq, regulons, use_aracne = T, ref = ref_R_Baseline_no_RUNX1m, ref_name = "R", treat = ref_OR_Baseline_no_RUNX1m, treat_name = "OR", minsize = 4, ges.filter = T)

R_OR_no_RUNX1m_msViper_combinatory <- msviperCombinatorial(R_OR_no_RUNX1m_msViper$mrs, regulators = 25, verbose = FALSE)
R_OR_no_RUNX1m_msViper_synergy <- msviperSynergy(R_OR_no_RUNX1m_msViper_combinatory, verbose = FALSE)

R_NR_no_RUNX1m_msViper <- run_msviper(RNAseq, regulons, use_aracne = T, ref = ref_R_Baseline_no_RUNX1m, ref_name = "R", treat = ref_NR_Baseline_no_RUNX1m, treat_name = "NR", minsize = 4, ges.filter = T)

R_NR_no_RUNX1m_msViper_combinatory <- msviperCombinatorial(R_NR_no_RUNX1m_msViper$mrs, regulators = 25, verbose = FALSE)
R_NR_no_RUNX1m_msViper_synergy <- msviperSynergy(R_NR_no_RUNX1m_msViper_combinatory, verbose = FALSE)
```



